### PR TITLE
[Validator] Add a `requireTld` option to `Url` constraint

### PR DESCRIPTION
**Pull request by javiereguiluz**
| Q             | A
| ------------- | ---
| Branch?       | 7.1
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Issues        | #50871
| License       | MIT

This implements the last part of #50871 following the suggestions from Wouter and Christian in
https://github.com/symfony/symfony/issues/50871#issuecomment-1871046803

### CODE REVIEW

File: src/Symfony/Component/Validator/Tests/Constraints/UrlValidatorTest.php [hunk 1]
```diff
@@ -311,6 +311,39 @@ public static function getValidCustomUrls()
...
+            $this->assertNoViolation();
+        } else {
+            $this->buildViolation($constraint->tldMessage)
+                ->setParameter('{{ value }}', '"'.$url.'"')
+                ->setCode(Url::MISSING_TLD_ERROR)
+                ->assertRaised();
+        }
+    }
+
+    public static function getUrlsForRequiredTld()
```
Comment (OskarStark) [inline src/Symfony/Component/Validator/Tests/Constraints/UrlValidatorTest.php]:
```suggestion
    public static function getUrlsForRequiredTld(): iterable
```

**Comment by xabbuh** [comment]
we need an entry in the Validator component's changelog file

File: src/Symfony/Component/Validator/Tests/Constraints/UrlValidatorTest.php (hunk 1 above)
Comment (javiereguiluz) [reply src/Symfony/Component/Validator/Tests/Constraints/UrlValidatorTest.php]: Added the return type. Thanks!

**Comment by javiereguiluz** [comment]
I've added the changelog entry. Thanks!

**Review by xabbuh — APPROVED** [review]

File: src/Symfony/Component/Validator/Constraints/UrlValidator.php [hunk 2]
```diff
@@ -64,6 +64,15 @@ public function validate(mixed $value, Constraint $constraint): void
             return;
         }
 
+        if ($constraint->requireTld && !str_contains($value, '.')) {
```
Comment (smnandre) [inline src/Symfony/Component/Validator/Constraints/UrlValidator.php]: Did that not open the door to the following "false-positive" ? 

- http://foo/bar.png
- http://user.pass@local
- http://127.0.0.1
- '../foo'

File: src/Symfony/Component/Validator/Constraints/UrlValidator.php [hunk 3]
```diff
@@ -64,6 +64,15 @@ public function validate(mixed $value, Constraint $constraint): void
...
+            $this->context->buildViolation($constraint->tldMessage)
+                ->setParameter('{{ value }}', $this->formatValue($value))
+                ->setCode(Url::MISSING_TLD_ERROR)
+                ->addViolation();
+
+            return;
+        }
+
```
Comment (smnandre) [inline src/Symfony/Component/Validator/Constraints/UrlValidator.php]: I'd move this _after_ the normalizer call

File: src/Symfony/Component/Validator/Constraints/UrlValidator.php (hunk 2 above)
Comment (javiereguiluz) [reply src/Symfony/Component/Validator/Constraints/UrlValidator.php]: Good catch! I've updated the validation logic. Thanks for the review!

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @javiereguiluz.

File: src/Symfony/Component/Validator/Constraints/Url.php [hunk 4]
```diff
@@ -48,13 +53,15 @@ public function __construct(
...
         mixed $payload = null,
+        ?bool $requireTld = null,
     ) {
         parent::__construct($options, $groups, $payload);
 
         $this->message = $message ?? $this->message;
         $this->protocols = $protocols ?? $this->protocols;
         $this->relativeProtocol = $relativeProtocol ?? $this->relativeProtocol;
         $this->normalizer = $normalizer ?? $this->normalizer;
+        $this->requireTld = $requireTld ?? $this->requireTld;
```
Comment (wouterj) [inline src/Symfony/Component/Validator/Constraints/Url.php:64]: In the issue, we discussed deprecating not setting this, so we can change the default to `true` (which is in sync with the `Hostname` constraint's `requireTld` option).

Do we still want to do this?

If we don't, I would say we change the default of the parameter to `false` instead (I see no point making it nullable?)

Comment (xabbuh) [reply src/Symfony/Component/Validator/Constraints/Url.php:64]: see #54588

**Pull request merged**